File: Next Greater Element II.py
#https://leetcode.com/problems/next-greater-element-ii/

class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        ans=[]
        stack=[]
        i=len(nums)-1
        #traversal
        while i>-1:
            while stack and stack[-1]<nums[i]:
                stack.pop()
            inserted=0
            if stack:
                if stack[-1]>nums[i]:
                    inserted=1
                    ans.insert(0,stack[-1])
                    stack.append(nums[i])
                else:
                    #non circular
                    for j in range(i+1,len(nums)):
                        if nums[j]>nums[i]:
                            inserted=1
                            ans.insert(0,nums[j])
                            #stack
                            while stack:
                                if stack[-1]<=nums[i]:
                                    stack.pop()
                                else:
                                    stack.append(nums[i])
                                    break
                            if len(stack)==0:
                                stack.append(nums[i])
                            break


                    #circular
                    if inserted!=1:
                        for j in range(0,i):
                            if nums[j]>nums[i]:
                                inserted=1
                                ans.insert(0,nums[j])
                                #stack
                                while stack:
                                    if stack[-1]<=nums[i]:
                                        stack.pop()
                                    else:
                                        stack.append(nums[i])
                                        break
                                if len(stack)==0:
                                    stack.append(nums[i])
                                break
            else:
                for j in range(0,i):
                    if nums[j]>nums[i]:
                        inserted=1
                        ans.insert(0,nums[j])
                        #stack
                        while stack:
                            if stack[-1]<=nums[j]:
                                stack.pop()
                            else:
                                stack.append(nums[i])
                                break
                        if len(stack)==0:
                            stack.append(nums[i])
                        break
            if inserted==0:
                ans.insert(0,-1)
                stack.append(nums[i])
            i-=1
        return ans


        

File: Next Greater Element II.cpp
//

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int l=nums.size();
        vector<int> result(l,-1);
        stack<int> bucket;
        for(int i=l-1;i>-1;i--){
            if(i==l-1){
                for(int j=l-1;j>-1;j--){// for last element... traverse once in reverse order... [2,5,4,7,2,1] for such case...
                    if(nums[j]>nums[i]){
                        bucket.push(nums[j]);
                        result[i]=nums[j];
                    }
                }
                bucket.push(nums[i]);
            }
            else{
                while(!bucket.empty()){
                    int t=bucket.top();
                    if(t<=nums[i]){
                        bucket.pop();
                    }
                    else{
                        result[i]=t;
                        break;
                    }
                }
                bucket.push(nums[i]);
            }
        }
        return result;
    }
};
